test/hotspot/jtreg/serviceability/jvmti/vthread/FollowReferences/libVThreadStackRefTest.cpp

```diff
@@ -0,0 +1,202 @@
+      }
+    }
+    break;
+    default:
+      // unexpected ref.kind
+      LOG("ERROR: unexpected ref_kind for class %d: %d\n",
+          (int)index, (int)reference_kind);
+    }
+  }
+  return JVMTI_VISIT_OBJECTS;
+}
+
+extern "C" JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM *vm, char *options, void *reserved) {
+  if (vm->GetEnv(reinterpret_cast<void **>(&jvmti), JVMTI_VERSION) != JNI_OK || jvmti == nullptr) {
+    LOG("Could not initialize JVMTI\n");
+    return JNI_ERR;
+  }
+  jvmtiCapabilities capabilities;
+  memset(&capabilities, 0, sizeof(capabilities));
+  capabilities.can_tag_objects = 1;
+  //capabilities.can_support_virtual_threads = 1;
+  jvmtiError err = jvmti->AddCapabilities(&capabilities);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("JVMTI AddCapabilities error: %d\n", err);
+    return JNI_ERR;
+  }
+
+  return JNI_OK;
+}
+
+
+/////////////////////////////////////////
+// Test native methods
+/////////////////////////////////////////
+extern "C" JNIEXPORT void JNICALL
+Java_VThreadStackRefTest_test(JNIEnv* env, jclass clazz, jobjectArray classes) {
+  jsize classesCount = env->GetArrayLength(classes);
+  for (int i=0; i<classesCount; i++) {
+    jvmti->SetTag(env->GetObjectArrayElement(classes, i), TAG_START + i);
+  }
+  refCounters.init(env, classesCount);
+  jvmtiHeapCallbacks heapCallBacks;
+  memset(&heapCallBacks, 0, sizeof(jvmtiHeapCallbacks));
+  heapCallBacks.heap_reference_callback = HeapReferenceCallback;
+  jvmtiError err = jvmti->FollowReferences(0, nullptr, nullptr, &heapCallBacks, nullptr);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("JVMTI FollowReferences error: %d\n", err);
+    env->FatalError("FollowReferences failed");
+  }
+}
+
+extern "C" JNIEXPORT jint JNICALL
+Java_VThreadStackRefTest_getRefCount(JNIEnv* env, jclass clazz, jint index) {
+  return refCounters.count[index];
+}
+
+extern "C" JNIEXPORT jlong JNICALL
+Java_VThreadStackRefTest_getRefThreadID(JNIEnv* env, jclass clazz, jint index) {
+  return refCounters.threadId[index];
+}
+
+static void printtCreatedClass(JNIEnv* env, jclass cls) {
+  jmethodID mid = env->GetMethodID(cls, "toString", "()Ljava/lang/String;");
+  if (mid == nullptr) {
+    env->FatalError("failed to get toString method");
+    return;
+  }
+  jstring jstr = (jstring)env->CallObjectMethod(cls, mid);
+  const char* str = env->GetStringUTFChars(jstr, 0);
+  LOG("created %s\n", str);
+  env->ReleaseStringUTFChars(jstr, str);
+}
+
+extern "C" JNIEXPORT void JNICALL
+Java_VThreadStackRefTest_createObjAndCallback(JNIEnv* env, jclass clazz, jclass cls, jobject callback) {
+  jobject jobj = env->AllocObject(cls);
+  printtCreatedClass(env, cls);
+
+  jclass callbackClass = env->GetObjectClass(callback);
+  jmethodID mid = env->GetMethodID(callbackClass, "run", "()V");
+  if (mid == nullptr) {
+    env->FatalError("cannot get run method");
+    return;
+  }
+  env->CallVoidMethod(callback, mid);
+}
+
+static volatile bool timeToExit = false;
```

Comment:
It is not enough to make variable volatile in c++. You need to make it atomic or use monitors to correctly synchronize.

Reply:
fixed